File: src/ai/.x/templates/openai-functions-streaming-go/function_call_context.go
package main

import (
	"fmt"

	"github.com/Azure/azure-sdk-for-go/sdk/ai/azopenai"
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
)

type FunctionCallContext struct {
	functionFactory   *FunctionFactory
	messages          []azopenai.ChatMessage
	functionName      string
	functionArguments string
}

func NewFunctionCallContext(functionFactory *FunctionFactory, messages []azopenai.ChatMessage) *FunctionCallContext {
	return &FunctionCallContext{
		functionFactory:   functionFactory,
		messages:          messages,
		functionName:      "",
		functionArguments: "",
	}
}

func (fcc *FunctionCallContext) CheckForUpdate(choice azopenai.ChatChoice) bool {
	updated := false

	if choice.Delta != nil && choice.Delta.FunctionCall != nil {
		name := choice.Delta.FunctionCall.Name
		if name != nil && *name != "" {
			fcc.functionName = *name
			updated = true
		}
	}

	if choice.Delta != nil && choice.Delta.FunctionCall != nil {
		args := choice.Delta.FunctionCall.Arguments
		if args != nil && *args != "" {
			fcc.functionArguments = *args
			updated = true
		}
	}

	return updated
}

func (fcc *FunctionCallContext) TryCallFunction() string {
	result := fcc.functionFactory.TryCallFunction(fcc.functionName, fcc.functionArguments)
	if result == "" {
		return ""
	}

	fmt.Printf("\rassistant-function: %s(%s) => %s\n", fcc.functionName, fcc.functionArguments, result)

	fcc.messages = append(fcc.messages, azopenai.ChatMessage{Role: to.Ptr(azopenai.ChatRoleAssistant), FunctionCall: &azopenai.ChatMessageFunctionCall{Name: to.Ptr(fcc.functionName), Arguments: to.Ptr(fcc.functionArguments)}})
	fcc.messages = append(fcc.messages, azopenai.ChatMessage{Role: to.Ptr(azopenai.ChatRoleFunction), Content: to.Ptr(result), Name: to.Ptr(fcc.functionName)})

	return result
}

func (fcc *FunctionCallContext) Clear() {
	fcc.functionName = ""
	fcc.functionArguments = ""
}

File: src/ai/.x/templates/openai-chat-webpage-with-functions-ts/src/FunctionCallContext.ts
import { ChatChoice, ChatRequestMessage } from "@azure/openai";
import { FunctionFactory } from "./FunctionFactory";

export class FunctionCallContext {
  private function_factory: FunctionFactory;
  private messages: ChatRequestMessage[];
  private function_name: string;
  private function_arguments: string;

  constructor(function_factory: FunctionFactory, messages: ChatRequestMessage[]) {
    this.function_factory = function_factory;
    this.messages = messages;
    this.function_name = "";
    this.function_arguments = "";
  }

  checkForUpdate(choice: ChatChoice): boolean {
    let updated = false;

    const name = choice.delta?.functionCall?.name;
    if (name !== undefined) {
      this.function_name = name;
      updated = true;
    }

    const args = choice.delta?.functionCall?.arguments;
    if (args !== undefined) {
      this.function_arguments = `${this.function_arguments}${args}`;
      updated = true;
    }

    return updated;
  }

  tryCallFunction(): string | undefined {
    let result = this.function_factory.tryCallFunction(this.function_name, this.function_arguments);
    if (result === undefined) {
      return undefined;
    }

    console.log(`assistant-function: ${this.function_name}(${this.function_arguments}) => ${result}`);

    this.messages.push({ role: 'assistant', content: '', functionCall: { name: this.function_name, arguments: this.function_arguments } });
    this.messages.push({ role: 'function', content: result, name: this.function_name });

    return result;
  }

  clear(): void {
    this.function_name = "";
    this.function_arguments = "";
  }
}
